### PR TITLE
python311Packages.qpsolvers: init at 3.4.0

### LINKED ISSUE #244480: Package request: qpsolvers
**Project description**

qpsolvers is python library for solving quadratic programming problems. It contains wrappers for multiple solvers like osqp, cvxopt, quadprog and much more.

**Metadata**

* homepage URL: https://github.com/qpsolvers/qpsolvers
* source URL: https://github.com/qpsolvers/qpsolvers
* license: LGPLv3
* platforms: platform independent (python3.10)

### PR DESCRIPTION
**Pull request by renesat**
###### Description of changes

qpsolvers is python library for solving quadratic programming problems. It contains wrappers for multiple solvers like osqp, cvxopt, quadprog and much more.

Metadata

    homepage URL: https://github.com/qpsolvers/qpsolvers
    source URL: https://github.com/qpsolvers/qpsolvers
    license: LGPLv3
    platforms: platform independent (python3.10)

Close #244480

Depend on #244528, #244529 and #244530

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/qpsolvers/default.nix [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  format = "pyproject";
+
+  src = fetchFromGitHub {
+    owner = "qpsolvers";
+    repo = "qpsolvers";
+    rev = "v${version}";
+    hash = "sha256-GrYAhTWABBvU6rGoHi00jBa7ryjCNgzO/hQBTdSW9cg=";
+  };
+
+  nativeBuildInputs = [flit unittestCheckHook];
```
Comment (natsukium) [inline pkgs/development/python-modules/qpsolvers/default.nix:28]:
```suggestion
```
ditto

File: pkgs/development/python-modules/qpsolvers/default.nix [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    scipy
+    scs
+
+    # For unittest
+    quadprog
+  ];
+
+  buildPhase = ''
+    flit build
+  '';
```
Comment (natsukium) [inline pkgs/development/python-modules/qpsolvers/default.nix:44]:
```suggestion
```

File: pkgs/development/python-modules/qpsolvers/default.nix [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
```
Comment (natsukium) [inline pkgs/development/python-modules/qpsolvers/default.nix:45]: Please add `pythonImportsCheck`.

File: pkgs/development/python-modules/qpsolvers/default.nix [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  meta = with lib; {
+    description = "Quadratic programming solvers in Python with a unified API";
+    homepage = "https://github.com/qpsolvers/qpsolvers";
+    license = licenses.lgpl3;
```
Comment (natsukium) [inline pkgs/development/python-modules/qpsolvers/default.nix:49]:
```suggestion
    license = licenses.lgpl3Plus;
```
https://github.com/qpsolvers/qpsolvers/blob/00545e02615a5efbfea4ba4dd3a324e2da77e4e0/qpsolvers/__init__.py#L10-L11

File: pkgs/development/python-modules/daqp/default.nix [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+  postPatch = ''
+    sed -i 's|../../../daqp|../..|' interfaces/daqp-python/setup.py
+  '';
+
+  buildPhase = ''
+    cd interfaces/daqp-python
+    python setup.py bdist_wheel
+  '';
+
```
Comment (natsukium) [inline pkgs/development/python-modules/daqp/default.nix:32]: Please add `pythonImportsCheck`.

**Review by natsukium — CHANGES_REQUESTED** [review]
Thanks for your contributions!

Would you please reformat with `nixpkgs-fmt`?

**Comment by renesat** [comment]
@natsukium Ok, all done.

File: pkgs/development/python-modules/daqp/default.nix [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  src = fetchFromGitHub {
+    owner = "darnstrom";
+    repo = "daqp";
+    rev = "5a15a3d16731d3d50f867218c1b281567db556fd";
+    hash = "sha256-in7Ci/wM7i0csJ4XVfo1lboWOyfuuU+8E+TzGmMV3x0=";
+  };
+
+  sourceRoot = "source/interfaces/daqp-python";
+
+  nativeBuildInputs = [ unittestCheckHook ];
```
Comment (natsukium) [inline pkgs/development/python-modules/daqp/default.nix:25]:
```suggestion
  nativeCheckInputs = [ unittestCheckHook ];
```

File: pkgs/development/python-modules/daqp/default.nix [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    wheel
+    numpy
+  ];
+
+  pythonImportsCheck = [ "daqp" ];
+
+  postPatch = ''
+    sed -i 's|../../../daqp|../..|' setup.py
+    sed -i 's|if src_path and os.path.exists(src_path):|if False:|' setup.py
+  '';
```
Comment (natsukium) [inline pkgs/development/python-modules/daqp/default.nix:39]: To make it easier to understand source code changes, could you please move `postPatch` right after `src`?

**Comment by renesat** [comment]
Fixed!

**Comment by natsukium** [comment]
Result of `nixpkgs-review pr 244532` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>32 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.bambi</li>
    <li>python310Packages.bambi.dist</li>
    <li>python310Packages.qiskit</li>
    <li>python310Packages.qiskit-aer</li>
    <li>python310Packages.qiskit-aer.dist</li>
    <li>python310Packages.qiskit-finance</li>
    <li>python310Packages.qiskit-finance.dist</li>
    <li>python310Packages.qiskit-ibmq-provider</li>
    <li>python310Packages.qiskit-ibmq-provider.dist</li>
    <li>python310Packages.qiskit-ignis</li>
    <li>python310Packages.qiskit-ignis.dist</li>
    <li>python310Packages.qiskit-nature</li>
    <li>python310Packages.qiskit-nature.dist</li>
    <li>python310Packages.qiskit-optimization</li>
    <li>python310Packages.qiskit-optimization.dist</li>
    <li>python310Packages.qiskit.dist</li>
    <li>python310Packages.scikit-survival</li>
    <li>python310Packages.scikit-survival.dist</li>
    <li>python311Packages.qiskit-aer</li>
    <li>python311Packages.qiskit-aer.dist</li>
    <li>python311Packages.qiskit-finance</li>
    <li>python311Packages.qiskit-finance.dist</li>
    <li>python311Packages.qiskit-ibmq-provider</li>
    <li>python311Packages.qiskit-ibmq-provider.dist</li>
    <li>python311Packages.qiskit-ignis</li>
    <li>python311Packages.qiskit-ignis.dist</li>
    <li>python311Packages.qiskit-nature</li>
    <li>python311Packages.qiskit-nature.dist</li>
    <li>python311Packages.qiskit-optimization</li>
    <li>python311Packages.qiskit-optimization.dist</li>
    <li>python311Packages.scikit-survival</li>
    <li>python311Packages.scikit-survival.dist</li>
  </ul>
</details>
<details>
  <summary>8 packages failed to build:</summary>
  <ul>
    <li>python310Packages.blackjax</li>
    <li>python310Packages.blackjax.dist</li>
    <li>python310Packages.jaxopt</li>
    <li>python310Packages.jaxopt.dist</li>
    <li>python311Packages.blackjax</li>
    <li>python311Packages.blackjax.dist</li>
    <li>python311Packages.jaxopt</li>
    <li>python311Packages.jaxopt.dist</li>
  </ul>
</details>
<details>
  <summary>24 packages built:</summary>
  <ul>
    <li>python310Packages.cvxpy</li>
    <li>python310Packages.cvxpy.dist</li>
    <li>python310Packages.daqp</li>
    <li>python310Packages.daqp.dist</li>
    <li>python310Packages.osqp</li>
    <li>python310Packages.osqp.dist</li>
    <li>python310Packages.qpsolvers</li>
    <li>python310Packages.qpsolvers.dist</li>
    <li>python310Packages.qutip</li>
    <li>python310Packages.qutip.dist</li>
    <li>python310Packages.scs</li>
    <li>python310Packages.scs.dist</li>
    <li>python311Packages.cvxpy</li>
    <li>python311Packages.cvxpy.dist</li>
    <li>python311Packages.daqp</li>
    <li>python311Packages.daqp.dist</li>
    <li>python311Packages.osqp</li>
    <li>python311Packages.osqp.dist</li>
    <li>python311Packages.qpsolvers</li>
    <li>python311Packages.qpsolvers.dist</li>
    <li>python311Packages.qutip</li>
    <li>python311Packages.qutip.dist</li>
    <li>python311Packages.scs</li>
    <li>python311Packages.scs.dist</li>
  </ul>
</details>

File: pkgs/development/python-modules/daqp/default.nix [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+  nativeCheckInputs = [ unittestCheckHook ];
+
+  propagatedBuildInputs = [
+    cython
+    setuptools
+    wheel
```
[the remark this answers is not in the document]
Comment (renesat) [reply pkgs/development/python-modules/daqp/default.nix]: Done + I fix unittest discover

**Review by natsukium — COMMENTED** [review]
LGTM, but the scs test seems to hang on x86_64-linux.🤔

**Comment by renesat** [comment]
> LGTM, but the scs test seems to hang on x86_64-linux.🤔

Hmmm... On my laptop all fine:
<details> 
  <summary>nix build .#python3Packages.scs </summary>
  <code>
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing setuptools-build-hook
Using setuptoolsBuildPhase
Using setuptoolsShellHook
Sourcing pip-install-hook
Using pipInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing setuptools-check-hook
Using setuptoolsCheckPhase
Sourcing pytest-check-hook
Using pytestCheckPhase
Removing setuptoolsCheckPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/wd5h7sgdalf727b6938zba0wawi38mvb-source
source root is source
setting SOURCE_DATE_EPOCH to timestamp 315619200 of file source/test/test_solve_random_cone_prob_mkl.py
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
no configure script, doing nothing
@nix { "action": "setPhase", "phase": "buildPhase" }
building
Executing setuptoolsBuildPhase
Namespace(scs=False, gpu=False, mkl=False, openmp=False, float32=False, extraverbose=False, gpu_atrans=True, int32=False, blas64=False)
/nix/store/b7rq6rjxyg23rwx0s8bv6vi7k6kwm3py-python3.10-setuptools-67.4.0/lib/python3.10/site-packages/setuptools/__init__.py:85: _DeprecatedInstaller: setuptools.installer and fetch_build_eggs are deprecated. Requirements should be satisfied by a PEP 517 installer. If you are using pip, you can try `pip install --use-pep517`.
  dist.fetch_build_eggs(dist.setup_requires)
running bdist_wheel
running build
running build_py
creating build
creating build/lib.linux-x86_64-cpython-310
creating build/lib.linux-x86_64-cpython-310/scs
copying src/__init__.py -> build/lib.linux-x86_64-cpython-310/scs
running build_ext
setup.py:97: DeprecationWarning: 

  `numpy.distutils` is deprecated since NumPy 1.23.0, as a result
  of the deprecation of `distutils` itself. It will be removed for
  Python >= 3.12. For older Python versions it will remain present.
  It is recommended to use `setuptools < 60.0` for those Python versions.
  For more details, see:
    https://numpy.org/devdocs/reference/distutils_status_migration.html 


  from numpy.distutils.system_info import get_info
blas_armpl_info:
  NOT AVAILABLE
blas_mkl_info:
  NOT AVAILABLE
blis_info:
  NOT AVAILABLE
openblas_info:
    libraries = ['lapack', 'lapacke', 'blas', 'cblas', 'lapack', 'lapacke', 'blas', 'cblas']
    library_dirs = ['/nix/store/amwml0fw056mnqviqf0ps7zc71kzyfx4-lapack-3/lib', '/nix/store/zs7hl4lk9y3wc2pixkincxyq6i35xd84-blas-3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    runtime_library_dirs = ['/nix/store/amwml0fw056mnqviqf0ps7zc71kzyfx4-lapack-3/lib', '/nix/store/zs7hl4lk9y3wc2pixkincxyq6i35xd84-blas-3/lib']
blas_opt_info:
    libraries = ['lapack', 'lapacke', 'blas', 'cblas', 'lapack', 'lapacke', 'blas', 'cblas']
    library_dirs = ['/nix/store/amwml0fw056mnqviqf0ps7zc71kzyfx4-lapack-3/lib', '/nix/store/zs7hl4lk9y3wc2pixkincxyq6i35xd84-blas-3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    runtime_library_dirs = ['/nix/store/amwml0fw056mnqviqf0ps7zc71kzyfx4-lapack-3/lib', '/nix/store/zs7hl4lk9y3wc2pixkincxyq6i35xd84-blas-3/lib']
lapack_armpl_info:
  NOT AVAILABLE
lapack_mkl_info:
  NOT AVAILABLE
openblas_lapack_info:
    libraries = ['lapack', 'lapacke', 'blas', 'cblas', 'lapack', 'lapacke', 'blas', 'cblas']
    library_dirs = ['/nix/store/amwml0fw056mnqviqf0ps7zc71kzyfx4-lapack-3/lib', '/nix/store/zs7hl4lk9y3wc2pixkincxyq6i35xd84-blas-3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    runtime_library_dirs = ['/nix/store/amwml0fw056mnqviqf0ps7zc71kzyfx4-lapack-3/lib', '/nix/store/zs7hl4lk9y3wc2pixkincxyq6i35xd84-blas-3/lib']
lapack_opt_info:
    libraries = ['lapack', 'lapacke', 'blas', 'cblas', 'lapack', 'lapacke', 'blas', 'cblas']
    library_dirs = ['/nix/store/amwml0fw056mnqviqf0ps7zc71kzyfx4-lapack-3/lib', '/nix/store/zs7hl4lk9y3wc2pixkincxyq6i35xd84-blas-3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
    runtime_library_dirs = ['/nix/store/amwml0fw056mnqviqf0ps7zc71kzyfx4-lapack-3/lib', '/nix/store/zs7hl4lk9y3wc2pixkincxyq6i35xd84-blas-3/lib']
Supported SIMD extensions in this NumPy install:
    baseline = SSE,SSE2,SSE3
    found = SSSE3,SSE41,POPCNT,SSE42,AVX,F16C,FMA3,AVX2
    not found = AVX512F,AVX512CD,AVX512_SKX,AVX512_CLX,AVX512_CNL,AVX512_ICL
Blas info:
{'libraries': ['lapack', 'lapacke', 'blas', 'cblas', 'lapack', 'lapacke', 'blas', 'cblas'], 'library_dirs': ['/nix/store/amwml0fw056mnqviqf0ps7zc71kzyfx4-lapack-3/lib', '/nix/store/zs7hl4lk9y3wc2pixkincxyq6i35xd84-blas-3/lib'], 'language': 'c', 'define_macros': [('HAVE_CBLAS', None)], 'runtime_library_dirs': ['/nix/store/amwml0fw056mnqviqf0ps7zc71kzyfx4-lapack-3/lib', '/nix/store/zs7hl4lk9y3wc2pixkincxyq6i35xd84-blas-3/lib']}
Lapack info:
{'libraries': ['lapack', 'lapacke', 'blas', 'cblas', 'lapack', 'lapacke', 'blas', 'cblas'], 'library_dirs': ['/nix/store/amwml0fw056mnqviqf0ps7zc71kzyfx4-lapack-3/lib', '/nix/store/zs7hl4lk9y3wc2pixkincxyq6i35xd84-blas-3/lib'], 'language': 'c', 'define_macros': [('HAVE_CBLAS', None)], 'runtime_library_dirs': ['/nix/store/amwml0fw056mnqviqf0ps7zc71kzyfx4-lapack-3/lib', '/nix/store/zs7hl4lk9y3wc2pixkincxyq6i35xd84-blas-3/lib']}
/nix/store/b7rq6rjxyg23rwx0s8bv6vi7k6kwm3py-python3.10-setuptools-67.4.0/lib/python3.10/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
  warnings.warn(
Finished executing setuptoolsBuildPhase
@nix { "action": "setPhase", "phase": "installPhase" }
installing
Executing pipInstallPhase
/build/source/dist /build/source
Processing ./scs-3.2.3-cp310-cp310-linux_x86_64.whl
Requirement already satisfied: numpy>=1.7 in /nix/store/3ypm1j8qa1prs2xincgxxh2cjr7cdbv6-python3.10-numpy-1.24.2/lib/python3.10/site-packages (from scs==3.2.3) (1.24.2)
Requirement already satisfied: scipy>=0.13.2 in /nix/store/l0jqa20zxqpmnf8s0yggvhgn54xyh23i-python3.10-scipy-1.10.1/lib/python3.10/site-packages (from scs==3.2.3) (1.10.1)
Installing collected packages: scs
Successfully installed scs-3.2.3
/build/source
Finished executing pipInstallPhase
@nix { "action": "setPhase", "phase": "pythonOutputDistPhase" }
pythonOutputDistPhase
Executing pythonOutputDistPhase
Finished executing pythonOutputDistPhase
@nix { "action": "setPhase", "phase": "fixupPhase" }
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/v58803xygn02h597445k4w6b50wyh7l4-python3.10-scs-3.2.3
shrinking /nix/store/v58803xygn02h597445k4w6b50wyh7l4-python3.10-scs-3.2.3/lib/python3.10/site-packages/_scs_direct.cpython-310-x86_64-linux-gnu.so
shrinking /nix/store/v58803xygn02h597445k4w6b50wyh7l4-python3.10-scs-3.2.3/lib/python3.10/site-packages/_scs_indirect.cpython-310-x86_64-linux-gnu.so
checking for references to /build/ in /nix/store/v58803xygn02h597445k4w6b50wyh7l4-python3.10-scs-3.2.3...
patching script interpreter paths in /nix/store/v58803xygn02h597445k4w6b50wyh7l4-python3.10-scs-3.2.3
stripping (with command strip and flags -S -p) in  /nix/store/v58803xygn02h597445k4w6b50wyh7l4-python3.10-scs-3.2.3/lib
shrinking RPATHs of ELF executables and libraries in /nix/store/4adwy4x82rwfpyf20qv59wd905z12cxg-python3.10-scs-3.2.3-dist
checking for references to /build/ in /nix/store/4adwy4x82rwfpyf20qv59wd905z12cxg-python3.10-scs-3.2.3-dist...
patching script interpreter paths in /nix/store/4adwy4x82rwfpyf20qv59wd905z12cxg-python3.10-scs-3.2.3-dist
Executing pythonRemoveTestsDir
Finished executing pythonRemoveTestsDir
@nix { "action": "setPhase", "phase": "installCheckPhase" }
running install tests
no Makefile or custom installCheckPhase, doing nothing
@nix { "action": "setPhase", "phase": "pythonCatchConflictsPhase" }
pythonCatchConflictsPhase
@nix { "action": "setPhase", "phase": "pythonRemoveBinBytecodePhase" }
pythonRemoveBinBytecodePhase
@nix { "action": "setPhase", "phase": "pythonImportsCheckPhase" }
pythonImportsCheckPhase
Executing pythonImportsCheckPhase
Check whether the following modules can be imported: scs
@nix { "action": "setPhase", "phase": "pytestCheckPhase" }
pytestCheckPhase
Executing pytestCheckPhase
[1m============================= test session starts ==============================[0m
platform linux -- Python 3.10.12, pytest-7.2.1, pluggy-1.0.0
rootdir: /build/source
[1mcollecting ... [0m[1m
collecting 5 items                                                             [0m[1m
collected 26 items                                                             [0m

test/test_scs_basic.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m                                             [ 19%][0m
test/test_scs_object.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m                                           [ 42%][0m
test/test_scs_rand.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m                                              [ 61%][0m
test/test_scs_sdp.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m                                               [ 80%][0m
test/test_solve_random_cone_prob.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m                                [100%][0m

[32m======================== [32m[1m26 passed[0m[32m in 107.20s (0:01:47)[0m[32m ========================[0m
Finished executing pytestCheckPhase
pytestCheckPhase completed in 1 minutes 54 seconds
@nix { "action": "setPhase", "phase": "pytestcachePhase" }
pytestcachePhase
@nix { "action": "setPhase", "phase": "pytestRemoveBytecodePhase" }
pytestRemoveBytecodePhase
</code>
</details>

**Comment by natsukium** [comment]
You are right. I can build scs with `nix-build -A python3Packages.scs`.
Let's disable that test only for x86_64-linux.

**Comment by renesat** [comment]
Done 👍

**Review by natsukium — APPROVED** [review]
Thanks! LGTM

**Pull request merged**